diff -uNr createrepo-0.9.9/modifyrepo.py createrepo-0.9.9p/modifyrepo.py
--- createrepo-0.9.9/modifyrepo.py	2013-04-11 17:12:21.527658715 -0500
+++ createrepo-0.9.9p/modifyrepo.py	2013-04-11 17:11:47.134443332 -0500
@@ -114,7 +114,7 @@
 
         do_compress = False
         ## Compress the metadata and move it into the repodata
-        if self.compress or not mdname.split('.')[-1] in ('gz', 'bz2', 'xz'):
+        if self.compress and not mdname.split('.')[-1] in ('gz', 'bz2', 'xz'):
             do_compress = True
             mdname += '.' + self.compress_type
         mdtype = self._get_mdtype(mdname, mdtype)
