diff -uNr oauth2-1.1.3/oauth2/__init__.py oauth2-1.1.3p/oauth2/__init__.py
--- oauth2-1.1.3/oauth2/__init__.py	2010-03-27 13:17:54.000000000 -0500
+++ oauth2-1.1.3p/oauth2/__init__.py	2011-01-14 16:24:29.000000000 -0600
@@ -26,6 +26,7 @@
 import time
 import random
 import urlparse
+import sha
 import hmac
 import binascii
 import httplib2
@@ -685,11 +686,6 @@
         key, raw = self.signing_base(request, consumer, token)

         # HMAC object.
-        try:
-            from hashlib import sha1 as sha
-        except ImportError:
-            import sha # Deprecated
-
         hashed = hmac.new(key, raw, sha)

         # Calculate the digest base 64.
