diff -Nur pyOpenSSL-0.5.1.orig/doc/tools/mkhowto pyOpenSSL-0.5.1/doc/tools/mkhowto
--- pyOpenSSL-0.5.1.orig/doc/tools/mkhowto	2001-07-30 13:31:48.000000000 +0200
+++ pyOpenSSL-0.5.1/doc/tools/mkhowto	2003-08-08 16:46:10.000000000 +0200
@@ -56,7 +56,7 @@
 DVIPS_BINARY = "dvips"
 LATEX_BINARY = "latex"
 LATEX2HTML_BINARY = "latex2html"
-LYNX_BINARY = "lynx"
+LYNX_BINARY = "elinks"
 MAKEINDEX_BINARY = "makeindex"
 PDFLATEX_BINARY = "pdflatex"
 PERL_BINARY = "perl"
@@ -397,7 +397,7 @@
         if tempdir is None:
             tempdir = self.doc
         indexfile = os.path.join(tempdir, "index.html")
-        self.run("%s -nolist -dump %s >%s.txt"
+        self.run("%s -dump %s >%s.txt"
                  % (LYNX_BINARY, indexfile, self.doc))
 
     def require_temps(self, binary=None):
diff -Nur pyOpenSSL-0.5.1.orig/setup.cfg pyOpenSSL-0.5.1/setup.cfg
--- pyOpenSSL-0.5.1.orig/setup.cfg	2002-09-10 00:05:21.000000000 +0200
+++ pyOpenSSL-0.5.1/setup.cfg	2003-08-08 16:48:34.000000000 +0200
@@ -4,7 +4,7 @@
 # bdist_rpm settings contributed by Mihai Ibanescu <misa@redhat.com>
 [bdist_rpm]
 release = 1
-build-requires = lynx openssl-devel python-devel perl tetex-dvips tetex-latex
+build-requires = elinks openssl-devel python-devel perl tetex-dvips tetex-latex
 group = Development/Libraries
 build_script = rpm/build_script
 doc-files = doc/pyOpenSSL.txt doc/pyOpenSSL.ps doc/html
