--- a/mod_wsgi/mod_wsgi.c 2012-02-01 10:31:51.791866441 -0500
+++ b/mod_wsgi/mod_wsgi.c	2012-02-01 10:33:02.738059228 -0500
@@ -14995,6 +14995,7 @@
 #endif
 
     static const char * const p6[] = { "mod_python.c", NULL };
+    static const char * const p7[] = { "mod_ssl.c", NULL };
 
     ap_hook_post_config(wsgi_hook_init, p6, NULL, APR_HOOK_MIDDLE);
     ap_hook_child_init(wsgi_hook_child_init, p6, NULL, APR_HOOK_MIDDLE);
@@ -15023,7 +15024,7 @@
     ap_register_provider(p, AUTHZ_PROVIDER_GROUP, "wsgi-group",
                          AUTHZ_PROVIDER_VERSION, &wsgi_authz_provider);
 #endif
-    ap_hook_access_checker(wsgi_hook_access_checker, NULL, n5, APR_HOOK_MIDDLE);
+    ap_hook_access_checker(wsgi_hook_access_checker, p7, n5, APR_HOOK_MIDDLE);
 #endif
 }
 
